dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -7,3263 +7,6479 @@
+#ifdef DDS_HAS_WCHAR
+  case TK_CHAR16: {
+    DDS::WcharSeq wcharseq;
+    return reconstruct_primitive_collection(wcharseq, size, bound, ACE_OutputCDR::from_wchar(0)) &&
+      (ser << wcharseq);
+  }
+#endif
+  case TK_BYTE: {
+    DDS::ByteSeq byteseq;
+    return reconstruct_primitive_collection(byteseq, size, bound, ACE_OutputCDR::from_octet(0x00)) &&
+      (ser << byteseq);
+  }
+  case TK_BOOLEAN: {
+    DDS::BooleanSeq boolseq;
+    return reconstruct_primitive_collection(boolseq, size, bound, ACE_OutputCDR::from_boolean(false)) &&
+      (ser << boolseq);
+  }
+  }
+  return false;
+}
+
+// Unlike primitive types, string and wstring are not as easy to reconstruct since each
+// element string may need to be reconstructed individually. Instead, we serialize
+// the element strings one by one directly.
+void DynamicDataImpl::DataContainer::serialized_size_string_common(const DCPS::Encoding& encoding,
+  size_t& size, const char* str) const
+{
+  primitive_serialized_size_ulong(encoding, size);
+  if (str) {
+    size += ACE_OS::strlen(str) + 1; // Include null termination
+  }
+}
+
+#ifdef DDS_HAS_WCHAR
+void DynamicDataImpl::DataContainer::serialized_size_string_common(const DCPS::Encoding& encoding,
+  size_t& size, const CORBA::WChar* wstr) const
+{
+  primitive_serialized_size_ulong(encoding, size);
+  if (wstr) {
+    size += ACE_OS::strlen(wstr) * DCPS::char16_cdr_size; // Not include null termination
+  }
+}
+#endif
+
+void DynamicDataImpl::DataContainer::serialized_size_string_common(const DCPS::Encoding& encoding,
+  size_t& size, const SingleValue& sv) const
+{
+  if (sv.kind_ == TK_STRING8) {
+    serialized_size_string_common(encoding, size, sv.str_);
+  }
+#ifdef DDS_HAS_WCHAR
+  else if (sv.kind_ == TK_STRING16) {
+    serialized_size_string_common(encoding, size, sv.wstr_);
+  }
+#endif
+}
+
+template<typename StringType>
+bool DynamicDataImpl::DataContainer::serialized_size_generic_string_collection(
+  const DCPS::Encoding& encoding, size_t& size, const IndexToIdMap& index_to_id) const
+{
+  for (CORBA::ULong i = 0; i < index_to_id.size(); ++i) {
+    const DDS::MemberId id = index_to_id[i];
+    if (id != MEMBER_ID_INVALID) {
+      const_single_iterator single_it = single_map_.find(id);
+      if (single_it != single_map_.end()) {
+        serialized_size_string_common(encoding, size, single_it->second);
+      } else if (!serialized_size_complex_member_i(encoding, size, id)) {
+        return false;
+      }
+    } else {
+      StringType default_value;
+      set_default_basic_value(default_value);
+      serialized_size_string_common(encoding, size, default_value);
+    }
+  }
+  return true;
+}
+
+template<typename StringType>
+bool DynamicDataImpl::DataContainer::serialized_size_generic_string_sequence(
+  const DCPS::Encoding& encoding, size_t& size, const IndexToIdMap& index_to_id) const
+{
+  if (encoding.xcdr_version() == DCPS::Encoding::XCDR_VERSION_2) {
+    serialized_size_delimiter(encoding, size);
+  }
```

Comment:
```suggestion
  serialized_size_delimiter(encoding, size);
```

These are redundant because the function already checks that.